packages/core/src/StateNode.ts

```diff
@@ -466,14 +485,25 @@ class StateNode<
    * @param context Custom context (will override predefined context)
    */
   public withConfig(
-    options: Partial<MachineOptions<TContext, TEvent>>,
+    options: Partial<MachineOptions<TContext, TEvent, TExtra>>,
     context?: TContext | (() => TContext)
-  ): StateNode<TContext, TStateSchema, TEvent, TTypestate> {
-    const { actions, activities, guards, services, delays } = this.options;
+  ): StateNode<TContext, TStateSchema, TEvent, TTypestate, TExtra> {
+    const {
+      actions,
+      activities,
+      guards,
+      services,
+      delays,
+      input
+    } = this.options;
 
     return new StateNode(
       this.config,
       {
+        input: {
+          ...input,
+          ...options.input
+        },
```

Comment:
I think I've mentioned this somewhere else too - it seems inconsistent how we approach input in this PR. In some cases, it's being overridden and in some cases, it's being merged. I know this is not unique to this PR though - similar things are inconsistent in more places of the codebase. However, I think it would still be nice to decide what's best for input and other things right now - so we don't have to change the input's behavior in the future.